### PR TITLE
Separate the DSL code and the more specific code

### PR DESCRIPTION
**Pull request by Lethalpaga**
This moves all the generic DSL code in a separate file so that it can be used by other means than the provided cli + calls to cfn_cmd.

Note that although I was careful to make it backward-compatible for the end-users. it will probably cause merge conflicts for all the other active branches in github

### CODE REVIEW

**Comment by jonaf** [comment]
Hi @Lethalpaga, can you describe some use-cases for separating the DSL? Do you intend to use the DSL syntax for something other than Cloudformation?

**Comment by temujin9** [comment]
@jonaf Actually, I can answer that. It's step 1 in reimplementing without the cfn_cmd dependency.

**Comment by temujin9** [comment]
(Okay, in fairness: that's what it made me excited for. His motives may be different.)

**Comment by temujin9** [comment]
Looks pretty good to us. We would certainly like to know your use cases for it, too.

Could you squash your commits (see [Contributing](https://github.com/bazaarvoice/cloudformation-ruby-dsl/blob/master/docs/Contributing.md#git) docs)?

**Comment by Lethalpaga** [comment]
All done :+1:  (I've made a few small bugfixes since the last commit)

My main problem with the way it worked before was that it's really only meant to be run as an executable. That doesn't work for me because given the way my CI works I have to generate the template in a Rakefile, and :
  - I don't like to shell out
  - I needed a way to feed parameters to the template without using ARGV
  - I'm only interested in the template JSON, the CI does the actual stack creation, so I needed a safe way to save the template to a file (getting output from *expand* causes problems if someone outputs debug traces in the template)

So basically I needed a way to use the template DSL as a module, and the tight integration with cfn_cmd made that hard (that and I get a bit psycho when useful code gets heavily coupled with something I don't care about :stuck_out_tongue_winking_eye:) .
